mediapipe/tasks/ios/components/containers/sources/MPPEmbedding.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 The MediaPipe Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/**
+ * Represents the embedding for a given embedder head. Typically used in embedding tasks.
+ *
+ * One and only one of the two 'floatEmbedding' and 'quantizedEmbedding' will contain data, based on
+ * whether or not the embedder was configured to perform scala quantization.
+ */
+NS_SWIFT_NAME(Embedding)
+@interface MPPEmbedding : NSObject
+
+/**
+ * @brief The Floating-point embedding.
+ * Empty if the embedder was configured to perform scalar quantization.
+ */
+@property(nonatomic, readonly, nullable) float *floatEmbedding;
+
+/**
+ * @brief The Quantized embedding.
+ * Empty if the embedder was not configured to perform scalar quantization.
+ */
+@property(nonatomic, readonly, nullable) char *quantizedEmbedding;
```

Comment:
@schmidt-sebastian @khanhlvg `quantizedEmbedding` is of type bytes in the proto. It is converted to `std::string` in C++. I have not used NSString* because if any operations are to be performed with the embedding, then the consumers may have to convert it into a char* array. Instead I am using char * directly as the type here. Moreover the embedding may be long since this is intended to use with image embedder as well. In case of streaming operations, continuous conversion to NSString may present extra over head.

Reply:
`NSData` is a wrapper class for any byte buffer.  It is not a primitive type unlike a `byte`  array in java. The primitive equivalent is `char *` like in C. 
Advantage of `NSData` is that for iOS devs it is a familiar type. Types like `char` `char *` etc. are not used in day to day tasks. 
Plus if initialised in the right way, it will take care of deallocating the underlying byte buffer. But In our case,  even if we choose `char *,` the task library will be handling deallocation of this property once `MPPEmbedding` goes out of scope. So that's not really a problem.
Disadvantage is that`NSData` in our case is an overhead for performance. Especially since we can expect embeddings to be used for some other task.  For this we will have to directly access and manipulate the underlying byte buffer.  So converting the `std::string` to `NSData` only for the consumers of the API to access the pointer seems costly. Also devs dealing with unconventional tasks like manipulation of embeddings will be expected to handle these kind of types and their deallocation if, they choose to copy. 

So the question is whether we should use `NSData` for the sake of familiarity?

(https://developer.apple.com/documentation/foundation/nsdata?language=objc)
I noticed that web is using a [`uint8 array`
](https://github.com/google/mediapipe/blob/master/mediapipe/tasks/web/components/containers/embedding_result.d.ts#L35). We can use int8* instead of char * if need be.

Reply:
`NSData` is a wrapper class for any byte buffer.  It is not a primitive type. Unlike `byte` array in java. Primitive equivalent is `char *` like in C. 

**Advantages of `NSData`**

- Familiar type for iOS devs. Types like `char *` are not used in regular tasks. 
- If initialised in the right way, it will take care of deallocating the underlying byte buffer. If we choose `char *,` the task library will be handling deallocation of this property once `MPPEmbedding` goes out of scope. So that's not really a problem.

**Disadvantage**
Overhead for performance. Especially since we can expect embeddings to be used for some other task.  For this we will have to directly access and manipulate the underlying byte buffer.  So converting the `std::string` to `NSData` only for the consumers of the API to access the pointer seems costly. 

Devs dealing with unconventional tasks like integrating embedding models will be expected to handle these kind of types and their deallocation if, they choose to copy. 
So the question is whether we should use `NSData` for the sake of familiarity?

I noticed that web is using a [`uint8 array`
](https://github.com/google/mediapipe/blob/master/mediapipe/tasks/web/components/containers/embedding_result.d.ts#L35). We can use `int8*` instead of `char *` since that makes more sense for quantized values.

Reply:
In the TFLite Task Lib image segmenter API, we used float array pointer directly to represent the segmentation masks only because it's more performant than using high-level data type like NSData or NSArray, and the segmentation masks are large enough to make those performance difference noticeable. However, embedding vector are usually small and I don't think there's a need to proceed embedding values independently like in segmentation mask, so using `NSData` sounds like a good idea to me.

Reply:
It looks like we are creating a copy of the mask even in the ImageSegmenter (tensorflow_lite_support/ios/task/processor/sources/TFLSegmentationResult.m). This avoids the lifetime issue and is a good choice here since the masks are float values. In the embedding case, since we are talking about bytes, I feel that we should just use NSData (and copy). This avoids lifetime issues and makes our API much more usable (otherwise, we would have to return a pointer and the length).

Reply:
I have updated `quantizedEmbedding` to be an `NSData`. Considering the above feedback, `floatEmbedding` has been changed to `NSArray<NSNumber *>*`. Note that to keep these types uniform, we can make them both `NSArray`s or `NSData`s if need be. LMK your thoughts.